pkg/imgutil/dockerconfigresolver/defaults.go

```diff
@@ -0,0 +1,51 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package dockerconfigresolver
+
+import "errors"
+
+type scheme string
+
+const (
+	standardHTTPSPort        = "443"
+	schemeHTTP        scheme = "http"
+	schemeHTTPS       scheme = "https"
+	// schemeNerdctlExperimental is currently provisional, to unlock namespace based host authentication
+	// This may change or break without notice, and you should have no expectations that credentials saved like that
+	// will be supported in the future
+	schemeNerdctlExperimental scheme = "nerdctl-experimental"
+	// See https://github.com/moby/moby/blob/v27.1.1/registry/config.go#L42-L48
+	//nolint:misspell
+	// especially Sebastiaan comments on future domain consolidation
+	dockerIndexServer = "https://index.docker.io/v1/"
+	// The query parameter that containerd will slap on namespaced hosts
+	namespaceQueryParameter = "ns"
+)
+
+// Errors returned by the credentials store
+var (
+	ErrUnableToInstantiate = errors.New("unable to instantiate docker credentials store")
+	ErrUnableToErase       = errors.New("unable to erase credentials")
+	ErrUnableToStore       = errors.New("unable to store credentials")
+	ErrUnableToRetrieve    = errors.New("unable to retrieve credentials")
+)
+
+// Errors returned by `Parse`
+var (
+	ErrUnparsableURL     = errors.New("unparsable registry URL")
+	ErrUnsupportedScheme = errors.New("unsupported scheme in registry URL")
+)
```

Comment:
nit: are they used widely in nerdctl or by downstream? if not I prefer just returning a `errors.New(...)` directly when needed instead of predefining them.

Reply:
> nit: are they used widely in nerdctl or by downstream?

Not yet.

The reason I like them is that they allow downstream code to actually test (`errors.Is(err, ErrXYZ)`) the error (instead of matching strings, which I find abhorrent...), and allows providing more refined user feedback based on these (or decide to ignore certain errors in certain circumstances).

This of course applies to testing as well (as a nice benefit, where we can actually verify what error is being returned in integration testing without having to duplicate strings again).

LMK if you feel strongly about these and of course open to amending.

Reply:
Overall motivation for me is that our error-ing is not great right now (see #3302 for example), and that is generally in the codebase.

We tend to just bubble-up system errors that are meaningless or even misleading for the user (#3197) - and at best, makes it hard to pinpoint for us when reading bug reports.

This is obviously a very large endeavor to change this situation overall, and it might take a few iterations to come up with a reasonable set of errors and informative enough wrapping, but I believe it is essential.